### PR TITLE
Add aggressive restrictions to pre-handshake p2p buffer limit

### PR DESCRIPTION
**Pull request by vtnerd**
I have tested this on my local node, but it could use some more testing before merge (this PR is to v0.17 branch). The buffer limit for p2p payloads is reduced to 256 KiB until a handshake is requested (outgoing) or successfully processed (incoming).

This places a tighter binding between the context and the levin code. Previously, the context was only required to have `boost::uuids::uuid m_connection_id` as a member.

### CODE REVIEW

File: src/p2p/net_node.h [hunk 1]
```diff
@@ -118,6 +118,8 @@ namespace nodetool
         m_in_timedsync(false)
     {}
 
+    static constexpr int handshake_command() noexcept { return 1001; }
```
Comment (moneromooo-monero) [inline src/p2p/net_node.h:121]: Can that be COMMAND_HANDSHAKE::ID ?

**Review by moneromooo-monero — COMMENTED** [review]
Should the reset to full packet size be in the notify code path ? If it is, it might be a way for a peer to make it look like a handshake was done, but not actually call the handshake handler (not sure whether it gets called in that case). Similarly, maybe it should be done only if the handler does not error out.

**Comment by vtnerd** [comment]
> Should the reset to full packet size be in the notify code path ? If it is, it might be a way for a peer to make it look like a handshake was done, but not actually call the handshake handler (not sure whether it gets called in that case). Similarly, maybe it should be done only if the handler does not error out.

The full packet size only gets sets in `async_invoke` and `invoke` when the local side is initiating the handshake, and after the `invoke` handler has processed a handshake and changed from `state_before_handshake`. I didn't use the return code because there are [positive return codes for errors](https://github.com/monero-project/monero/blob/master/src/p2p/net_node.inl#L2488) and for success. I didn't consider changing the return codes.

File: src/p2p/net_node.h (hunk 1 above)
Comment (vtnerd) [reply src/p2p/net_node.h:121]: `COMMAND_HANDSHAKE` is a `typedef` for `COMMAND_HANDSHAKE_T<t_payload_net_handler::payload_type>` inside of `node_server`, and the type `t_payload_net_handler` is a template argument to `node_server`. This is outside of the `node_server` class entirely, so I decided to just hard-code and add a `static_assert`.

Instead, I will do `{ return COMMAND_HANDSHAKE_T<int>::ID; }` with a comment, and the `static_assert` to ensure no specializations.

Comment (vtnerd) [reply src/p2p/net_node.h:121]: Or wait - is it just easier to leave as `{ return 1001; }` with the `static_assert`? Its unfortunate, but catches mistakes at build time.

**Comment by vtnerd** [comment]
> The full packet size only gets sets in async_invoke and invoke when the local side is initiating the handshake, and after the invoke handler has processed a handshake and changed from state_before_handshake. I didn't use the return code because there are positive return codes for errors and for success. I didn't consider changing the return codes.

And to clarify further - incoming and outgoing notify messages go to different code paths.

File: src/p2p/net_node.h (hunk 1 above)
Comment (moneromooo-monero) [reply src/p2p/net_node.h:121]: Fine by me to leave as is if's not that simple as just replacing.

**Review by moneromooo-monero — APPROVED** [review]

**Comment by binaryFate** [comment]
Reviewed from security perspective.

**Pull request merged**